### PR TITLE
Feature/upload result manually

### PR DESCRIPTION
**Pull request by lorenzoPrimi**
Fixes https://github.com/ooni/probe/issues/838
TestSummaryTableViewController.m became TestSummaryViewController.m but for some reason git think is a new file

### CODE REVIEW

**Comment by bassosimone** [comment]
This pull request is huge. I can either rubber stamp it or spend days to review it.

**Comment by bassosimone** [comment]
> This pull request is huge. I can either rubber stamp it or spend days to review it.

To further explain myself: PRs should be 200-300m lines, otherwise the review is painful. This will allow me to actually review it and give you more meaningful comments on the code, so it helps both of us :-).

File: ooniprobe/View/TestResults/TestResultsViewController.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#import <UIKit/UIKit.h>
```
Comment (lorenzoPrimi) [inline ooniprobe/View/TestResults/TestResultsViewController.h]: This file is not new, is the previous one who has been renamed

File: ooniprobe/View/TestResults/TestSummaryViewController.m [hunk 2]
```diff
@@ -0,0 +1,142 @@
+#import "TestSummaryViewController.h"
```
Comment (lorenzoPrimi) [inline ooniprobe/View/TestResults/TestSummaryViewController.m]: This file is not new, is the previous one who has been renamed

File: ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m [hunk 3]
```diff
@@ -31,4 +37,10 @@ - (void)viewDidLoad {
     [self.runtimeDetailLabel setTextColor:[UIColor colorWithRGBHexString:color_gray9 alpha:1.0f]];
 }
 
+-(IBAction)openMethodology:(id)sender{
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: @lorenzoPrimi is it correct to say that this `openMethodology` part is unrelated with uploads and instead is the UI feature that allows to open the methodology of a specific nettest?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 4]
```diff
@@ -0,0 +1,124 @@
+#import "UploadFooterViewController.h"
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: This file makes me a little bit sad. It seems that it contains both UI handling and code for resubmitting measurements. I would be happier if we could have a clean separation of these two concerns.

I'd like the code for resubmitting to be totally independent from UI, so that we can (1) submit it as a separate PR without wondering about the storyboard and (2) write some tests for it. The latter will allow us to be more confident that the code that we're merging into master works as intended.

File: ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m (hunk 3 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: Exactly. Bus as the storyboard was touched heavily in both branches I had to merge the methodology branch here. I couldn't be merged in master.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 4 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: The downside of having two classes is not having full control over UI events. As the upload is used only in the footer and the footer is a reusable class in any view it's best to keep most of the code here. The only function that can be moved away is `uploadSingleMeasurement` bu it will decrease the readability

Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: So basically this is not a real UI class but a detachable piece of UI

File: ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m (hunk 3 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: So, it seems the bottleneck is really the storyboard. I'm starting to wonder whether there are smart solutions around always hitting the storyboard and or best practices. _/me reads stuff_

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: So, [there is people suggesting to break the storyboard so that each view controller has its own](https://medium.com/practical-ios-development/how-real-teams-deal-with-ios-storyboards-beb8bb4c2765).

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: [Here they compare the pros and cons of storyboard and XIB](https://stackoverflow.com/questions/9404471/when-to-use-storyboard-and-when-to-use-xibs).

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: My point is the following: we should ask ourself the question of whether, if when we try to do sane coding the storyboard gets in the way, maybe it's not our fault, maybe the storyboard is the problem.

Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: Xib is so 2013 ! 
I used the approach to group storyboard in categories, like results.storyboard, settings.storyboard etc.. in this case is really difficult to have conflicts unless you really modify the same group of screens. 
Usually it shouldn't happen in fact I tend to add only one feature per version, in this case the methodology was heavily pushed and I should have waited this pr to be done to develop methodology

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: > I used the approach to group storyboard in categories
Right, this seems a good way to handle the problem.

> [...] I should have waited this pr to be done to develop methodology
I guess we can work on as many non UI features as we want, but only one UI feature at a time per storyboard can be in development at any single time. Does this make sense?

Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: Exactly. Also I plan to divide the Result storyboard into two in the future. My approach is to keep max 7-8 screens inside a storyboad. Unless they are striclty connected for preview reasons.

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/TestDetailsFooterViewController.m]: Ok, sounds reasonable!

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 4 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Okay, I'll need to think a little bit. My main concern is that I want some pieces of code to be testable and so I need to understand how to proceed here.

Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Eventually the upload functions are detachable, but
- There will be much code in general for the back and forth communication with UI
- One more class to manage uploaded used by only one class
- The `UploadFooterViewController` class will become only its `viewDidLoad` method

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I really want to pursue this goal of separating. If not now, maybe at a later time.

**Comment by bassosimone** [comment]
@lorenzoPrimi please, address the remaining comments. Then, I'll do a final pass.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+#import "MessageUtility.h"
+#import <QuartzCore/QuartzCore.h>
+#import "SettingsUtility.h"
+#import <mkall/MKCollector.h>
+#import "TestUtility.h"
+#import "MBProgressHUD.h"
+
+@interface UploadFooterViewController ()
+
+@end
```
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: This is the xcode generated file. I never touch it.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#import <UIKit/UIKit.h>
+#import "Result.h"
+#import "Measurement.h"
+
+@interface UploadFooterViewController : UIViewController
```
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.h]: done

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+-(void)showModalHelp{
+    UIAlertAction* okButton = [UIAlertAction
+                               actionWithTitle:NSLocalizedString(@"Modal.ResultsNotUploaded.Button.Upload", nil)
+                               style:UIAlertActionStyleDefault
+                               handler:^(UIAlertAction * action) {
+                                   [self uploadResult];
+                               }];
+    [MessageUtility alertWithTitle:NSLocalizedString(@"Modal.ResultsNotUploaded.Title", nil)
+                           message:NSLocalizedString(@"Modal.ResultsNotUploaded.Paragraph", nil)
+                          okButton:okButton
```
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: My `+ (void)alertWithTitle:(NSString *)title message:(NSString *)msg okButton:(UIAlertAction*)okButton inView:(UIViewController *)view` function auto add a cancel action who dismiss the modal

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+@interface UploadFooterViewController ()
+
+@end
+
+@implementation UploadFooterViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    if (_upload_all){
```
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: done

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I feel brave today :-). Can you please remove it? I'm sure I've written ObjectiveC code where this internal interface was not there, so we can safely remove it.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I see, thanks!

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Thanks!

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Removing it still builds the project.
The fact is this interface is in basically *every* file so maybe is better to have a PR with a general cleanup rather than doing it only once in one file

File: ooniprobe/View/TestResults/Details/TestDetailsViewController.m [hunk 9]
```diff
@@ -82,6 +99,12 @@ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
         [self.measurement setReRun];
         [vc setTestSuite:testSuite];
     }
+    else if ([[segue identifier] isEqualToString:@"footer_upload"]){
+        UploadFooterViewController *vc = (UploadFooterViewController * )segue.destinationViewController;
+        [vc setResult:result];
+        [vc setMeasurement:measurement];
+        [vc setUpload_all:false];
```
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Details/TestDetailsViewController.m]: @lorenzoPrimi can you please reply here? I do not fully understand the API you've written and it's not documented so I don't 100% grasp it. Can you add documentation to the new header file?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.h]: Please document also the properties explaining their meaning

File: ooniprobe/View/TestResults/Details/TestDetailsViewController.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Details/TestDetailsViewController.m]: upload_all is a bool I use to indicate if the footer should upload a measurement (one) or a result (all)

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.h]: It's sufficient to have a single doc sentence before each line explaining the purpose

[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.h]: I'm gonna document only upload_all as the others result, measurement, titleLabel are self explanatory, what do you think?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Let's start with not increasing the entropy in this PR. Refactoring other files can be done incrementally.

("Always leave the camp cleaner than you find it.")

[the remark this answers is not in the document]
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: so do you think this can be refactored here or in a next PR?

File: ooniprobe/View/TestResults/Details/TestDetailsViewController.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Details/TestDetailsViewController.m]: Thanks, I suspected that, but seeing it documented is better for everyone!

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.h]: I think a short piece of doc for every public attribute is better. Tip: if you use `///` rather than `//` for documentation comments, Xcode recognised that and provides more help.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Just remove the three lines in this file, so we don't ever commit them. We'll deal with the rest with time and calm as there is no rush to convert all the code base today, just the desired to make this PR minimal.
